spinalcordtoolbox/scripts/sct_deepseg.py

```diff
@@ -322,7 +332,12 @@ def main(argv: Sequence[str]):
                 keep_largest=1 if arguments.task == 'spinalcord' else arguments.keep_largest,
                 fill_holes_in_pred=arguments.fill_holes,
                 remove_small=arguments.remove_small,
-                use_gpu=use_gpu, remove_temp_files=arguments.r)
+                use_gpu=use_gpu, remove_temp_files=arguments.r,
+                # Pass any "extra" kwargs defined in task-specific subparsers
+                extra_inference_kwargs={arg_name: getattr(arguments, arg_name)
+                                        for arg_name in ["step1_only"]  # Used only by totalspineseg
+                                        if hasattr(arguments, arg_name)}
```

Comment:
Why is this a dict comprehension? `getattr` allows you to set a default return (as, under the hood, its just a dict query anyways):

```suggestion
                extra_inference_kwargs={"step1_only": getattr(arguments, "step1_only", None)}
```

*NOTE:* despite the default of `getattr`'s 'default' already being `None`, we need to specify it anyways due to how `getattr` works under the hood (raising an error on use if no default was explicitly define).

Reply:
I think my reasoning was that specifying `{'step1_only': None}` by default would cause an error for MONAI/nnUNet models? (Since the `step1_only` parameter is only present in the `segment_totalspineseg` function.)

```python
>>> from spinalcordtoolbox.deepseg.inference import segment_totalspineseg
>>> segment_totalspineseg("test", "test", "test", "test", **{'step1_only': None})
AssertionError: No 'nnUNet/results/rYYYYMMDD' subdirectory found in test

>>> from spinalcordtoolbox.deepseg.inference import segment_nnunet
>>> segment_nnunet("test", "test", "test", "test", **{'step1_only': None})
Traceback (most recent call last):
  File "/opt/pycharm-2024.2.3/plugins/python-ce/helpers/pydev/pydevconsole.py", line 364, in runcode
    coro = func()
  File "<input>", line 1, in <module>
TypeError: segment_nnunet() got an unexpected keyword argument 'step1_only'
```

The dictionary comprehension was just to default to `{}` if non-totalspineseg tasks are used. 

(Also, I figured the pattern was general enough that we could in theory add other subparser-specific arguments to the same list to be able to pass those along too.)

Reply:
Fair enough; at that point, though, would it not be smarter to simply bundle everything in `kwargs` wholesale and pass it through?

Reply:
Good question! I touched on this a little in the extended commit message:

> To accomplish this, we use a dictionary of "extra" keyword arguments to smuggle in kwargs that may be present for some tasks/frameworks/subparsers, but not for others.
>
> It's possible that this is a smell that the "common function signature" has outgrown its usefulness, and that we need to refactor how arguments are parsed for `sct_deepseg`? But this works for a first draft, and I like that it clearly delineates between common arguments and unique arguments.

In other words, I wanted to still try and enforce a common function signature with similar args (i.e. every NN framework function should have an input image, a predictor, a `torch` device, and a `tmpdir` to write outputs to). And the "extra kwags" parameter was a nod to the fact that `step1_only` is an outlier, and specific to `totalspineseg`.

If we combined everything into `kwargs`, then it increases the risk that some parameter could be mis-specified, isn't as self-documenting, etc. I'm kind of treating `kwargs` like a last resort, aha.